// electron-builder/out/codeSign.js
"use strict";

const util_1 = require("./util");
const fs_extra_p_1 = require("fs-extra-p");
const httpRequest_1 = require("./httpRequest");
const os_1 = require("os");
const path = require("path");
const promise_1 = require("./promise");
const bluebird_1 = require("bluebird");
const crypto_1 = require("crypto");
//noinspection JSUnusedLocalSymbols
const __awaiter = require("./awaiter");
function randomString() {
    return crypto_1.randomBytes(8).toString("hex");
}
function generateKeychainName() {
    return "csc-" + randomString() + ".keychain";
}
exports.generateKeychainName = generateKeychainName;
function createKeychain(keychainName, cscLink, cscKeyPassword, csaLink) {
    const authorityCertPath = path.join(os_1.tmpdir(), randomString() + ".cer");
    const developerCertPath = path.join(os_1.tmpdir(), randomString() + ".p12");
    const keychainPassword = randomString();
    return promise_1.executeFinally(bluebird_1.Promise.all([httpRequest_1.download(csaLink || "https://developer.apple.com/certificationauthority/AppleWWDRCA.cer", authorityCertPath), httpRequest_1.download(cscLink, developerCertPath), bluebird_1.Promise.mapSeries([["create-keychain", "-p", keychainPassword, keychainName], ["unlock-keychain", "-p", keychainPassword, keychainName], ["set-keychain-settings", "-t", "3600", "-u", keychainName]], it => util_1.exec("security", it))]).then(() => importCerts(keychainName, authorityCertPath, developerCertPath, cscKeyPassword)), errorOccurred => {
        const tasks = [fs_extra_p_1.deleteFile(authorityCertPath, true), fs_extra_p_1.deleteFile(developerCertPath, true)];
        if (errorOccurred) {
            tasks.push(deleteKeychain(keychainName));
        }
        return promise_1.all(tasks);
    });
}
exports.createKeychain = createKeychain;
function importCerts(keychainName, authorityCertPath, developerCertPath, cscKeyPassword) {
    return __awaiter(this, void 0, void 0, function* () {
        yield util_1.exec("security", ["import", authorityCertPath, "-k", keychainName, "-T", "/usr/bin/codesign"]);
        yield util_1.exec("security", ["import", developerCertPath, "-k", keychainName, "-T", "/usr/bin/codesign", "-P", cscKeyPassword]);
        let cscName = yield extractCommonName(cscKeyPassword, developerCertPath);
        return {
            cscName: cscName,
            cscKeychainName: keychainName
        };
    });
}
function extractCommonName(password, certPath) {
    return util_1.exec("openssl", ["pkcs12", "-nokeys", "-nodes", "-passin", "pass:" + password, "-nomacver", "-clcerts", "-in", certPath]).then(result => {
        const match = result[0].toString().match(/^subject.*\/CN=([^\/\n]+)/m);
        if (match == null || match[1] == null) {
            throw new Error("Cannot extract common name from p12");
        } else {
            return match[1];
        }
    });
}
function sign(path, options) {
    const args = ["--deep", "--force", "--sign", options.cscName, path];
    if (options.cscKeychainName != null) {
        args.push("--keychain", options.cscKeychainName);
    }
    return util_1.exec("codesign", args);
}
exports.sign = sign;
function deleteKeychain(keychainName) {
    let ignoreNotFound = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

    const result = util_1.exec("security", ["delete-keychain", keychainName]);
    if (ignoreNotFound) {
        return result.catch(error => {
            if (!(error.message.indexOf("The specified keychain could not be found.") !== -1)) {
                throw error;
            }
        });
    } else {
        return result;
    }
}
exports.deleteKeychain = deleteKeychain;
function downloadCertificate(cscLink) {
    const certPath = path.join(os_1.tmpdir(), randomString() + ".p12");
    return httpRequest_1.download(cscLink, certPath).thenReturn(certPath);
}
exports.downloadCertificate = downloadCertificate;
