// electron-builder/out/packager.js
"use strict";

const fs_1 = require("fs");
const path = require("path");
const util_1 = require("./util");
const promise_1 = require("./promise");
const events_1 = require("events");
const bluebird_1 = require("bluebird");
const metadata_1 = require("./metadata");
const platformPackager_1 = require("./platformPackager");
const errorMessages = require("./errorMessages");
const util = require("util");
const deepAssign = require("deep-assign");
//noinspection JSUnusedLocalSymbols
const __awaiter = require("./awaiter");
function addHandler(emitter, event, handler) {
    emitter.on(event, handler);
}
class Packager {
    //noinspection JSUnusedGlobalSymbols
    constructor(options) {
        let repositoryInfo = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];

        this.options = options;
        this.repositoryInfo = repositoryInfo;
        this.isTwoPackageJsonProjectLayoutUsed = true;
        this.eventEmitter = new events_1.EventEmitter();
        this.projectDir = options.projectDir == null ? process.cwd() : path.resolve(options.projectDir);
        this.appDir = this.computeAppDirectory();
    }
    artifactCreated(handler) {
        addHandler(this.eventEmitter, "artifactCreated", handler);
        return this;
    }
    get devPackageFile() {
        return path.join(this.projectDir, "package.json");
    }
    build() {
        return __awaiter(this, void 0, void 0, function* () {
            const devPackageFile = this.devPackageFile;
            const appPackageFile = this.projectDir === this.appDir ? devPackageFile : path.join(this.appDir, "package.json");
            const platforms = this.options.platform;
            const metadataList = yield bluebird_1.Promise.map(Array.from(new Set([devPackageFile, appPackageFile])), util_1.readPackageJson);
            this.metadata = metadataList[metadataList.length - 1];
            this.devMetadata = deepAssign(metadataList[0], this.options.devMetadata);
            this.checkMetadata(appPackageFile, devPackageFile, platforms);
            this.electronVersion = yield util_1.getElectronVersion(this.devMetadata, devPackageFile);
            const cleanupTasks = [];
            return promise_1.executeFinally(this.doBuild(platforms, cleanupTasks), () => promise_1.all(cleanupTasks.map(it => it())));
        });
    }
    doBuild(platforms, cleanupTasks) {
        return __awaiter(this, void 0, void 0, function* () {
            const distTasks = [];
            const outDir = path.resolve(this.projectDir, platformPackager_1.use(this.devMetadata.directories, it => it.output) || "dist");
            for (let platform of platforms) {
                const helper = this.createHelper(platform, cleanupTasks);
                for (let arch of normalizeArchs(platform, this.options.arch)) {
                    yield this.installAppDependencies(arch);
                    // electron-packager uses productName in the directory name
                    const appOutDir = path.join(outDir, `${ helper.appName }-${ platform.nodeName }-${ arch }`);
                    yield helper.pack(outDir, appOutDir, arch);
                    if (this.options.dist) {
                        distTasks.push(helper.packageInDistributableFormat(outDir, appOutDir, arch));
                    }
                }
            }
            return yield bluebird_1.Promise.all(distTasks);
        });
    }
    createHelper(platform, cleanupTasks) {
        if (this.options.platformPackagerFactory != null) {
            return this.options.platformPackagerFactory(this, platform, cleanupTasks);
        }
        switch (platform) {
            case metadata_1.Platform.OSX:
                {
                    const helperClass = require("./macPackager").default;
                    return new helperClass(this, cleanupTasks);
                }
            case metadata_1.Platform.WINDOWS:
                {
                    const helperClass = require("./winPackager").WinPackager;
                    return new helperClass(this, cleanupTasks);
                }
            case metadata_1.Platform.LINUX:
                return new (require("./linuxPackager").LinuxPackager)(this);
            default:
                throw new Error("Unknown platform: " + platform);
        }
    }
    // Auto-detect app/ (two package.json project layout (development and app)) or fallback to use working directory if not explicitly specified
    computeAppDirectory() {
        let customAppPath = this.options.appDir;
        let required = true;
        if (customAppPath == null) {
            customAppPath = util_1.DEFAULT_APP_DIR_NAME;
            required = false;
        }
        const absoluteAppPath = path.join(this.projectDir, customAppPath);
        try {
            fs_1.accessSync(absoluteAppPath);
        } catch (e) {
            if (required) {
                throw new Error(customAppPath + " doesn't exists, " + e.message);
            } else {
                this.isTwoPackageJsonProjectLayoutUsed = false;
                return this.projectDir;
            }
        }
        return absoluteAppPath;
    }
    checkMetadata(appPackageFile, devAppPackageFile, platforms) {
        const reportError = missedFieldName => {
            throw new Error("Please specify '" + missedFieldName + "' in the application package.json ('" + appPackageFile + "')");
        };
        const appMetadata = this.metadata;
        if (appMetadata.name == null) {
            reportError("name");
        } else if (appMetadata.description == null) {
            reportError("description");
        } else if (appMetadata.version == null) {
            reportError("version");
        } else if (appMetadata !== this.devMetadata && appMetadata.build != null) {
            throw new Error(util.format(errorMessages.buildInAppSpecified, appPackageFile, devAppPackageFile));
        } else if (this.devMetadata.build == null) {
            throw new Error(util.format(errorMessages.buildIsMissed, devAppPackageFile));
        } else {
            const author = appMetadata.author;
            if (author == null) {
                reportError("author");
            } else if (this.options.dist && author.email == null && platforms.indexOf(metadata_1.Platform.LINUX) !== -1) {
                throw new Error(util.format(errorMessages.authorEmailIsMissed, appPackageFile));
            }
        }
    }
    installAppDependencies(arch) {
        if (this.isTwoPackageJsonProjectLayoutUsed) {
            return util_1.installDependencies(this.appDir, this.electronVersion, arch, "rebuild");
        } else {
            util_1.log("Skipping app dependencies installation because dev and app dependencies are not separated");
            return bluebird_1.Promise.resolve();
        }
    }
}
exports.Packager = Packager;
function normalizeArchs(platform, arch) {
    if (platform === metadata_1.Platform.OSX) {
        return ["x64"];
    } else {
        return arch == null ? [process.arch] : arch === "all" ? ["ia32", "x64"] : [arch];
    }
}
exports.normalizeArchs = normalizeArchs;
function normalizePlatforms(platforms) {
    if (platforms == null || platforms.length === 0) {
        return [metadata_1.Platform.fromString(process.platform)];
    } else if (platforms[0] === "all") {
        if (process.platform === metadata_1.Platform.OSX.nodeName) {
            return [metadata_1.Platform.OSX, metadata_1.Platform.LINUX, metadata_1.Platform.WINDOWS];
        } else if (process.platform === metadata_1.Platform.LINUX.nodeName) {
            // OS X code sign works only on OS X
            return [metadata_1.Platform.LINUX, metadata_1.Platform.WINDOWS];
        } else {
            return [metadata_1.Platform.WINDOWS];
        }
    } else {
        return platforms.map(it => it instanceof metadata_1.Platform ? it : metadata_1.Platform.fromString(it));
    }
}
exports.normalizePlatforms = normalizePlatforms;
